### PR TITLE
Metadata writer refactoring

### PR DESCRIPTION
**Pull request by tmat**
Refactors PeWriter into components independent of Roslyn, so that it can later be moved to System.Reflection.Metadata in CoreFX.

The goal is to provide PE/COFF builder and ECMA-335 metadata builder that can be used broadly to write managed and native PE files. The goal is not to immediately provide the full support for writing all existing sections of the PE/COFF format but rather to enable composing PE/COFF files from section blobs and to provide builders for sections commonly appearing in managed PE files.

More work will follow to polish the public surface of the builders and serializers, as well as provide various ECMA-335 blob writers (to serialize signatures, attribute values, etc.).

### CODE REVIEW

**Comment by tmat** [comment]
test win please

**Comment by tmat** [comment]
@dotnet/roslyn-compiler Please review.
@nguerrera @ericfe-ms FYI

**Comment by tmat** [comment]
@dotnet/roslyn-compiler Ping.

**Comment by tmat** [comment]
@AlekseyTs @VSadov @jaredpar @agocke @gafter @TyOverby Anybody?

File: src/Compilers/Core/Portable/PEWriter/Members.cs [hunk 1]
```diff
@@ -184,7 +184,7 @@ ImmutableArray<byte> MarshallingDescriptor
         /// <summary>
         /// Offset of the field.
         /// </summary>
-        uint Offset
+        int Offset
         {
             get;
             // ^ requires this.ContainingTypeDefinition.Layout == LayoutKind.Explicit;
```
Comment (VSadov) [inline src/Compilers/Core/Portable/PEWriter/Members.cs]: i think we should remove these "^ requires", possibly changing some to asserts. Could be a separate change.

Comment (tmat) [inline src/Compilers/Core/Portable/PEWriter/Members.cs]: Agreed. There is also some unnecessary redundancy on these interfaces that can be eliminated.

**Comment by tmat** [comment]
Superseded by PR #7683, which adds signature encoding on top of this change.

**Pull request closed without merging**